### PR TITLE
[Tensor refactor] Updates the TF interface to exclude the quantum gradients for input tensors not watched by the GradientTape

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

This is the fourth PR of the variable refactor, following #652. This PR focuses on the TF interface so that watched variables of existing gradient tapes are used to determine the differentiability of QNode arguments.

**Description of the Change:**

* The TF interface now checks for non-differentiable positional arguments, and, if present, excludes them from the call to `JacobianQNode.jacobian()`. The VJP is then expanded, with 0s placed wrt the non-differentiable parameters.

**Benefits:**

* Non-differentiable parameters, when passed as positional arguments, are now _excluded_ from the parameter shift rule. Differentiable parameters are any parameters that are being watched by the gradient tape; including autowatched variables.

* More intuitive when working with data embedding and chained QNodes

**Possible Drawbacks:**

* The QNode parameter logic/signature introspection remains, so only positional arguments remain differentiable in the TF interface. Once the tensor refactor is complete, we can return to the QNode and see what can be deleted. If we are able to make keyword arguments differentiable in `JacobianQNode.jacobian`, we can extend all interfaces likewise.

* The relatively undocumented C++ Python wrapper `tensorflow.python.eager.tape.should_record_backprop` function is being used to check if QNode input tensors are being watched, as we do not have access to the users gradient tape(s). This function is not well documented, and only used in a single place in the TF code base, so could change/break between TF versions.

**Related GitHub Issues:** #648, #650, #613, #652

### CODE REVIEW

File: pennylane/interfaces/tf.py [hunk 1]
```diff
@@ -98,10 +101,20 @@ def __repr__(self):
...
         func = qnode.func
+        num_variables = property(lambda self: qnode.num_variables)
+        arg_vars = property(lambda self: qnode.arg_vars)
 
     @qnode_str
     @tf.custom_gradient
     def _TFQNode(*input_, **input_kwargs):
+        # determine which input tensors are being recorded for backpropagation
+        requires_grad = [
+            tape_lib.should_record_backprop([1 * i])
```
Comment (co9olguy) [inline pennylane/interfaces/tf.py:112]: Why is the `1` needed? :thinking:

Comment (co9olguy) [inline pennylane/interfaces/tf.py:112]: What is the return type of this method? Boolean?

File: pennylane/interfaces/tf.py [hunk 2]
```diff
@@ -125,9 +138,21 @@ def _TFQNode(*input_, **input_kwargs):
...
+                    indices = [i.idx for i in _flatten(arg_variable)]
+                    non_diff_indices.update(indices)
+
+            if non_diff_indices:
+                diff_indices = set(range(qnode.num_variables)) - non_diff_indices
+
             # evaluate the Jacobian matrix of the QNode
             variables = tfkwargs.get("variables", None)
-            jacobian = qnode.jacobian(args, kwargs)
+            jacobian = qnode.jacobian(args, kwargs, wrt=diff_indices)
```
Comment (co9olguy) [inline pennylane/interfaces/tf.py:155]: :pray:

File: tests/interfaces/test_tf.py [hunk 3]
```diff
@@ -768,3 +768,379 @@ def test_model_nested_tensor(self):
...
+
+        # input data
+        data1 = tf.constant([0, 1, 1, 0], dtype=tf.float64) / np.sqrt(2)
+        data2 = tf.Variable([1, 1], dtype=tf.float64, trainable=False)
+
+        with tf.GradientTape() as tape:
+            loss = circuit(weights, data1, data2)
+
+        grad = tape.gradient(loss, weights)
+        res = grad.numpy()
```
Comment (co9olguy) [inline tests/interfaces/test_tf.py:808]: How come we need to convert to numpy? We can't check using tensors?

File: tests/interfaces/test_tf.py [hunk 4]
```diff
@@ -768,3 +768,379 @@ def test_model_nested_tensor(self):
...
+
+        # we do not check for correctness, just that the output
+        # is the correct shape
+        assert res.shape == weights.shape
+
+        # check that the parameter shift was only performed for the
+        # differentiable elements of `weights`, not the data input
```
Comment (co9olguy) [inline tests/interfaces/test_tf.py:815]:
```suggestion
        # check that the gradient was computed for the
        # differentiable elements of `weights`, not the data input
```

File: tests/interfaces/test_tf.py [hunk 5]
```diff
@@ -768,3 +768,379 @@ def test_model_nested_tensor(self):
...
```
Comment (co9olguy) [inline tests/interfaces/test_tf.py:856]:
```suggestion
        # check that the gradient was computed for the
        # differentiable elements of `weights`, not the data input
```

File: tests/interfaces/test_tf.py [hunk 6]
```diff
@@ -768,3 +768,379 @@ def test_model_nested_tensor(self):
...
```
Comment (co9olguy) [inline tests/interfaces/test_tf.py:899]:
```suggestion
        # check that the gradient was computed for the
        # differentiable elements of `weights`, not the data input
```

File: tests/interfaces/test_tf.py [hunk 7]
```diff
@@ -768,3 +768,379 @@ def test_model_nested_tensor(self):
...
+        res1 = grad[0].numpy()
+        res2 = grad[1].numpy()
+
+        # we do not check for correctness, just that the output
+        # is the correct shape
+        assert res1.shape == weights1.shape
+        assert res2.shape == weights2.shape
+
+        # check that the parameter shift was only performed for the
+        # differentiable elements of `weights`, not the data input
```
Comment (co9olguy) [inline tests/interfaces/test_tf.py:944]:
```suggestion
        # check that the gradient was only computed for the
        # differentiable elements of `weights`, not the data input
```

File: tests/interfaces/test_tf.py [hunk 8]
```diff
@@ -768,3 +768,379 @@ def test_model_nested_tensor(self):
...
+            return qml.expval(qml.PauliZ(0))
+
+        weights = tf.constant(qml.init.strong_ent_layers_normal(n_wires=2, n_layers=3))
+
+        with tf.GradientTape() as tape:
+            tape.watch([weights])
+            loss = circuit(weights)
+
+        grad = tape.gradient(loss, weights)
+        assert grad is not None
```
Comment (co9olguy) [inline tests/interfaces/test_tf.py:991]: Cool

File: tests/interfaces/test_tf.py [hunk 9]
```diff
@@ -768,3 +768,379 @@ def test_model_nested_tensor(self):
...
+
+        expected = [
+            tf.cos(theta) * tf.sin(phi) * tf.sin(varphi),
+            tf.sin(theta) * tf.cos(phi) * tf.sin(varphi),
+        ]
+
+        assert np.allclose(grad[:2], expected, atol=tol, rtol=0)
+        assert grad[2] is None
+
+        # check that the parameter-shift rule was not applied to varphi
```
Comment (co9olguy) [inline tests/interfaces/test_tf.py:1070]:
```suggestion
        # check that the gradient was not computed for varphi
```

File: tests/interfaces/test_tf.py [hunk 10]
```diff
@@ -768,3 +768,379 @@ def test_model_nested_tensor(self):
...
+        # np.hstack 'flattens' the ragged gradient array allowing it
+        # to be compared with the expected result
+        assert np.allclose(np.hstack(grad), expected, atol=tol, rtol=0)
+
+        # Check that the parameter-shift rule was applied
+        # to all parameters in circuit2 (i.e., wrt=None)
+        assert spy.call_args_list[0][1]["wrt"] == None
+
+        # check that the parameter-shift rule was not applied
+        # to the first parameter of circuit1
```
Comment (co9olguy) [inline tests/interfaces/test_tf.py:1145]:
```suggestion
        # check that the gradient was not computed
        # for the first parameter of circuit1
```

File: tests/interfaces/test_tf.py [hunk 11]
```diff
@@ -768,3 +768,379 @@ def test_model_nested_tensor(self):
...
+            # analytic expression for dcost/dw[1]
+            0
+        ]
+
+        # np.hstack 'flattens' the ragged gradient array allowing it
+        # to be compared with the expected result
+        assert np.allclose(np.hstack(grad), expected, atol=tol, rtol=0)
+
+        # Check that the parameter-shift rule was applied
+        # to all parameters in circuit2 (i.e., wrt=None)
```
Comment (co9olguy) [inline tests/interfaces/test_tf.py:1141]:
```suggestion
        # Check that the gradient was computed
        # for all parameters in circuit2 (i.e., wrt=None)
```

File: tests/interfaces/test_tf.py [hunk 12]
```diff
@@ -768,3 +768,379 @@ def test_model_nested_tensor(self):
...
+        assert spy.call_args_list[0][1]["wrt"] == None
+
+        # check that the parameter-shift rule was not applied
+        # to the first parameter of circuit1
+        assert spy.call_args_list[1][1]["wrt"] == {1, 2}
```
Comment (co9olguy) [inline tests/interfaces/test_tf.py:1146]: Since these tests were copied from the two parallel PRs, a lot of the suggestions that I had in those ones carry over here (specifically writing the asserts out directly)

File: pennylane/interfaces/tf.py (hunk 1 above)
Comment (josh146) [reply pennylane/interfaces/tf.py:112]: Yep! Will expand my code comment to include this.

Comment (josh146) [reply pennylane/interfaces/tf.py:112]: It seems that this function only accepts tensors, not Variables. Generally, this shouldn't be an issue; `Variable` subclasses from `tf.core.Tensor`, so via duck typing all TF functions/classes that accept tensors should also work with Variables. But maybe this function is too low-level/close to C++ for duck typing to work?

Regardless, multiplying a Variable by a literal is a quick way of converting it into a tensor without breaking the computational graph, so that it will be accepted by this method.

I suppose it's clearer/more explicit to simply do `tf.convert_to_tensor(i)`.
```suggestion
            tape_lib.should_record_backprop([tf.convert_to_tensor(i)])
```

File: pennylane/interfaces/tf.py [hunk 13]
```diff
@@ -98,10 +101,20 @@ def __repr__(self):
         metric_tensor = qnode.metric_tensor
         draw = qnode.draw
         func = qnode.func
+        num_variables = property(lambda self: qnode.num_variables)
+        arg_vars = property(lambda self: qnode.arg_vars)
 
     @qnode_str
     @tf.custom_gradient
     def _TFQNode(*input_, **input_kwargs):
+        # determine which input tensors are being recorded for backpropagation
```
Comment (josh146) [inline pennylane/interfaces/tf.py:110]:
```suggestion
        # Determine which input tensors are being recorded for backpropagation.
        # The function should_record_backprop, documented here:
        # https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/eager/tape.py#L163
        # accepts lists of *tensors* (not Variables), returning True if all are being watched by one or more
        # existing gradient tape, False if not.
```

File: tests/interfaces/test_tf.py (hunk 8 above)
Comment (josh146) [reply tests/interfaces/test_tf.py:991]: This was something I declared in the ADR was 'not possible', since it's not documented anywhere in the TF API 😆

File: .github/CHANGELOG.md [hunk 14]
```diff
@@ -52,7 +54,12 @@
 
   - The QNode Autograd interface now inspects QNode positional arguments when calculating
     the vector-Jacobian product. If any argument is marked as `requires_grad=False`, it
-    is automatically excluded from the parameter-shift rule.
+    is automatically excluded from quantum gradient computations.
+
+  - The QNode TF interface now inspects QNode positional arguments when calculating
+    the vector-Jacobian product. If any argument is not being watched by a `tf.GradientTape()`,
+    it is automatically excluded from quantum gradient computations.
+    is automatically excluded from quantum gradient computations.
```
Comment (josh146) [inline .github/CHANGELOG.md:62]:
```suggestion
```

File: tests/interfaces/test_tf.py (hunk 12 above)
Comment (josh146) [reply tests/interfaces/test_tf.py:1137]: Have copied over the explicit asserts 👍

File: tests/interfaces/test_tf.py (hunk 3 above)
Comment (josh146) [reply tests/interfaces/test_tf.py]: Good catch; I've removed the superfluous numpy conversions.

File: pennylane/operation.py [hunk 15]
```diff
@@ -152,7 +152,7 @@ class ObservableReturnTypes(Enum):
 
     def __repr__(self):
         """String representation of the return types."""
-        return self.value
+        return str(self.value)
```
Comment (josh146) [inline pennylane/operation.py:155]: Codefactor was complaining about this not being a string. I'm assuming there has been a new pylint release, and this is wasn't being caught before.

File: pennylane/interfaces/tf.py [hunk 16]
```diff
@@ -98,10 +101,24 @@ def __repr__(self):
...
+        # The function should_record_backprop, documented here:
+        # https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/eager/tape.py#L163
+        # accepts lists of *tensors* (not Variables), returning True if all are being watched by one or more
+        # existing gradient tape, False if not.
+        requires_grad = [
+            tape_lib.should_record_backprop([tf.convert_to_tensor(i)])
+            if isinstance(i, (Variable, tf.Tensor))
+            else False
+            for i in input_
+        ]
```
Comment (trbromley) [inline pennylane/interfaces/tf.py:120]: I think I'm getting confused here, why do we care only about the input arguments and not keyword arguments?

File: pennylane/interfaces/tf.py [hunk 17]
```diff
@@ -136,6 +165,14 @@ def grad(grad_output, **tfkwargs):
...
             grad_input = tf.matmul(grad_output_row, jacobian)
             grad_input = tf.reshape(grad_input, [-1])
+
+            if non_diff_indices:
+                # TensorFlow requires we return a gradient of size (num_variables,)
+                res = np.zeros([qnode.num_variables])
+                indices = np.fromiter(diff_indices, dtype=np.int64)
+                res[indices] = grad_input
+                grad_input = tf.constant(res, dtype=dtype)
+
```
Comment (trbromley) [inline pennylane/interfaces/tf.py:175]: Nice, the process is actually quite similar to the torch version!

File: pennylane/interfaces/tf.py (hunk 16 above)
Comment (trbromley) [reply pennylane/interfaces/tf.py:120]: > The QNode parameter logic/signature introspection remains, so only positional arguments remain differentiable in the TF interface. Once the tensor refactor is complete, we can return to the QNode and see what can be deleted. If we are able to make keyword arguments differentiable in JacobianQNode.jacobian, we can extend all interfaces likewise.

I should read the PR description :laughing:
